stake-pool/single-pool-cli/src/cli.rs

```diff
@@ -0,0 +1,399 @@
+    #[clap(long = "stake-authority", value_parser = Pubkey::from_str)]
+    pub stake_authority_address: Option<Pubkey>,
+}
+
+#[derive(Clone, Debug, Args)]
+#[clap(group(ArgGroup::new("pool-source").required(true).args(&["pool-address", "vote-account-address"])))]
+pub struct DisplayCli {
+    /// The pool to display
+    #[clap(value_parser = Pubkey::from_str)]
+    pub pool_address: Option<Pubkey>,
+
+    /// The vote account corresponding to the pool to display
+    #[clap(long = "vote-account", value_parser = Pubkey::from_str)]
+    pub vote_account_address: Option<Pubkey>,
+}
+
+pub fn parse_json_rpc_url(url_or_moniker: &str) -> Result<String, String> {
+    is_url_or_moniker(url_or_moniker).map(|_| normalize_to_url_if_moniker(url_or_moniker))
+}
+
+pub fn parse_output_format(output_format: &str) -> OutputFormatArg {
+    match output_format {
+        "json" => OutputFormatArg::Json,
+        "json-compact" => OutputFormatArg::JsonCompact,
+        _ => unreachable!(),
+    }
+}
+
+pub fn is_valid_token_name(s: &str) -> Result<(), String> {
+    if s.len() > 32 {
+        Err("Maximum token name length is 32 characters".to_string())
+    } else {
+        Ok(())
+    }
+}
+
+pub fn is_valid_token_uri(s: &str) -> Result<(), String> {
+    if s.len() > 10 {
+        Err("Maximum token symbol length is 10 characters".to_string())
+    } else {
+        Ok(())
+    }
+}
+
+pub fn pool_address_from_args(maybe_pool: Option<Pubkey>, maybe_vote: Option<Pubkey>) -> Pubkey {
+    if let Some(pool_address) = maybe_pool {
+        pool_address
+    } else if let Some(vote_account_address) = maybe_vote {
+        find_pool_address(&single_pool::id(), &vote_account_address)
+    } else {
+        unreachable!()
+    }
+}
+
+// this plus with_signer(s) is because solana clap v3 utils requires ArgMatches
+// which means its impossible to parse keypairs in value_parser
+// instead, we take the input into a string wrapper from the cli
+// and then once the first pass is over, we do a second manual pass converting to signer wrappers
+#[derive(Clone, Debug)]
+pub enum SignerArg {
+    Source(String),
+    Signer(Arc<dyn Signer>),
+}
+impl FromStr for SignerArg {
+    type Err = std::convert::Infallible;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(Self::Source(s.to_string()))
+    }
+}
+impl PartialEq for SignerArg {
+    fn eq(&self, other: &SignerArg) -> bool {
+        match (self, other) {
+            (SignerArg::Source(ref a), SignerArg::Source(ref b)) => a == b,
+            (SignerArg::Signer(ref a), SignerArg::Signer(ref b)) => a == b,
+            (_, _) => false,
+        }
+    }
+}
+
+pub fn signer_from_arg(
+    signer_arg: Option<SignerArg>,
+    default_signer: &Arc<dyn Signer>,
+) -> Arc<dyn Signer> {
+    match signer_arg {
+        Some(SignerArg::Signer(signer)) => signer,
+        _ => default_signer.clone(),
+    }
+}
```

Comment:
Looking later, I think I understand this behavior more. Just to be sure though, is the default only meant to be used in the case of `None`? If so, can you return an error if it's `Some(SignerArg::Source(_))`?